blog/2019-09-ngrx-data-views/README.md

```diff
@@ -0,0 +1,419 @@
+
+As the `getBook` selector only provides access to IDs of authors and tags, we have to compose additional selectors for the display of these entities. The one for the authors would look like this:
+
+```ts
+export const getAuthorsOfBook = createSelector(
+  getBook,
+  getAuthorEntities,
+  (book, authors): Author[] =>
+    book
+      ? book.authorIds.map(authorId => authors[authorId])
+      : []
+)
+```
+
+Have a look at how we used the previous selector `getBook` in the new selector. When using a parameterized selector in another selector, the new one also becomes a parameterized selector.
+
+The `getAuthorsOfBook` selector will be used in the component like this:
+
+```ts
+authors$ = this.store.pipe(select(getAuthorsOfBook, this.bookId));
+```
+
+Similarly a selector to retrieve the tags of the current book has to be composed.
+
+At the end we will have the following properties in our component file:
+
+```ts
+book$ = this.store.pipe(select(getBook, this.bookId));
+authors$ = this.store.pipe(select(getAuthorsOfBook, this.bookId));
+tags$ = this.store.pipe(select(getTagsOfBook, this.bookId));
+```
+
+And we will use them on the template like this:
+
+```html
+<div *ngIf="book$ | async as book">
+  <div>{{ book.title }}</div>
+  <div><span *ngFor="let a of (authors$ | async)">{{ a.lastName }}</span></div>
+  <div>{{ book.description }}</div>
+  <div><span *ngFor="let t of (tags$ | async)">{{ t.name }}</span></div>
+</div>
+```
+
+## The Problem
+
+You can probably see what this will lead to in the long run: Code duplication. Every component displaying books will have to select the normalized data from the store and accumulate it on the template. Now imagine an API change. With this amount of duplication it could mean that every components typescript file as well as all Angular templates would have to be touched. But how can we solve this?
+
+Abstractly speaking: Currently we retrieve the data normalized via the API (1) and put it normalized into the state (2). Then we use selectors (3) to retrieve the data normalized in each component and only then de-normalize it on each and every template (4) that displays it.
+
+![Normalization points](./normalization_points.svg)
+
+It seems there are three points where we could intervene and do the de-normalization instead:
+
+1. Retrieve the data ready for display:
+  
+   This can only be done by changing from a simple REST based API to an API that provides access to all data with one call. GraphQL would be an answer, but most often this API is not provided.
+
+2. Put the data de-normalized into the store:
+  
+   This obviously leads to data duplication and problems when updating the data. It's simply not what we want.
+
+3. Select the data ready-for-display from the store:
+
+   This way the data is normalized in the store and updates can be done in an efficient way but there have to be special selectors that simplify selecting all accumulated data. - Let's build one!
```

Comment:
Emphasize that this is the cleanest solution. The store is still a normalized database, the "query" returns the exact data needed, as in real databases.